### DISCUSSION TITLE
Window losing focus when child button is pressed

### DISCUSSION THREAD

**Discussion by aekobear** [fltk-rs/fltk-rs#1131] [Q&A] [answered]
I've made a simple application with a window and two buttons. I want the window to close when it loses focus to other applications. However, clicking the second button within the window causes the window's unfocus event to trigger. And I cannot figure out any way to detect that the window is losing focus to a child component. Is this a bug? Am I doing something wrong?

```rust
use fltk::{app::{App}, prelude::*, button::Button, window::Window, enums::{Event}};

fn main() {
    let app = App::default();
    let mut window = Window::new(0, 0, 200, 200, "Example");

    window.handle(move |widget, ev| -> bool {
        match ev {
            Event::Unfocus => { println!("unfocused!"); app.quit(); true }
            _e => { false }
        }
    });

    let mut button_one = Button::new(0, 0, 200, 20, "click one");
    button_one.set_callback(move |_| {
        println!("clicked one");
    });

    let mut button_two = Button::new(0, 20, 200, 20, "click two");
    button_two.set_callback(move |_| {
        println!("clicked two");
    });

    window.end();
    window.show();
    app.run().unwrap();
}
```

**Comment by MoAlyousef** [comment, marked as answer]
I can't seem to repro with this code. Do you have a clear repro example?
Also what version of fltk-rs are you using, and what operating system?

Also you want to handle events correctly in the handle method.
true for handled or ignored events. false for unhandled events.

```rust
        match ev {
            Event::Unfocus => {
                println!("unfocused!");
                app.quit();
                true
            }
            _e => { false },          // unhandled
        }
```

**Reply by aekobear** [reply]
Sorry, I rushed with my example and missed a critical detail- you need two buttons for this to happen. The first button has focus automatically, so clicking the second button is what triggers the unfocus event on the window. I have updated my original question with a complete working reproduction

**Reply by MoAlyousef** [reply]
It seems FLTK sends an unfocus event while it changes focus between widgets. A partial solution is also to detect Event leave and enter (won't be triggered if another widget has focus):
```rust
use fltk::{app::App, button::Button, enums::Event, prelude::*, window::Window};

fn main() {
    let app = App::default();
    let mut window = Window::new(0, 0, 200, 200, "Example");
    window.handle({
        let mut left = false;
        move |widget, ev| -> bool {
            match ev {
                Event::Leave => {
                    left = true;
                    true
                }
                Event::Enter => {
                    left = false;
                    true
                }
                Event::Unfocus => {
                    if left {
                        app.quit();
                    }
                    true
                }
                _e => false,
            }
        }
    });

    let mut button_one = Button::new(0, 0, 200, 20, "click one");
    button_one.set_callback(move |_| {
        println!("clicked one");
    });

    let mut button_two = Button::new(0, 20, 200, 20, "click two");
    button_two.set_callback(move |_| {
        println!("clicked two");
    });

    window.end();
    window.show();

    app.run().unwrap();
}
```
A complete solution would probably require checking whether any widget in your app holds focus before quitting the app.

**Reply by aekobear** [reply]
I tried the Enter/Leave approach and interestingly, both entering and leaving seem to trigger the `Leave` event. `Enter` event is never triggered. This seems like a bug for sure.

Keeping track of focus in every button should work in theory, but I'm struggling to create a reference that can be shared among the static lifetime callbacks (still new to rust and learning the ropes).

Regardless, I am happy to mark this as "answered" unless you want to dig into the Enter/Leave bug more?

**Reply by MoAlyousef** [reply]
It seems to be a bug on Windows. I’ll repro in C++ and report upstream.

**Reply by aekobear** [reply]
Thanks!
